realm/realm-library/src/main/java/io/realm/OrderedRealmCollectionImpl.java

```diff
@@ -0,0 +1,604 @@
+     *         {@code false} otherwise
+     */
+    @Override
+    public boolean contains(Object object) {
+        if (isLoaded()) {
+            // Deleted objects can never be part of a RealmResults
+            if (object instanceof RealmObjectProxy) {
+                RealmObjectProxy proxy = (RealmObjectProxy) object;
+                if (proxy.realmGet$proxyState().getRow$realm() == InvalidRow.INSTANCE) {
+                    return false;
+                }
+            }
+
+            for (E e : this) {
+                if (e.equals(object)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns the element at the specified location in this list.
+     *
+     * @param location the index of the element to return.
+     * @return the element at the specified index.
+     * @throws IndexOutOfBoundsException if {@code location < 0 || location >= size()}.
+     */
+    @Override
+    public E get(int location) {
+        realm.checkIfValid();
+        return realm.get(classSpec, className, collection.getUncheckedRow(location));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E first() {
+        return firstImpl(true, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E first(E defaultValue) {
+        return firstImpl(false, defaultValue);
+    }
+
+    private E firstImpl(boolean shouldThrow, E defaultValue) {
+        UncheckedRow row = collection.firstUncheckedRow();
+
+        if (row != null) {
+            return realm.get(classSpec, className, row);
+        } else {
+            if (shouldThrow) {
+                throw new IndexOutOfBoundsException("No results were found.");
+            } else {
+                return defaultValue;
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E last() {
+        return lastImpl(true, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E last(E defaultValue) {
+        return lastImpl(false, defaultValue);
+
+    }
+
+    private E lastImpl(boolean shouldThrow, E defaultValue) {
+        UncheckedRow row = collection.lastUncheckedRow();
+
+        if (row != null) {
+            return realm.get(classSpec, className, row);
+        } else {
+            if (shouldThrow) {
+                throw new IndexOutOfBoundsException("No results were found.");
+            } else {
+                return defaultValue;
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deleteFromRealm(int location) {
+        // TODO: Implement the deleteLast in OS level and do check there!
+        realm.checkIfValidAndInTransaction();
+        collection.delete(location);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean deleteAllFromRealm() {
+        realm.checkIfValid();
+        if (size() > 0) {
```

Comment:
This is the current behavior https://github.com/realm/realm-java/blob/e318dca510b807495550820643fa5c0938e13acd/realm/realm-library/src/main/java/io/realm/RealmResults.java#L304